# About

Here is my work for the review of the client's idea.

I used the original dataset supplied by the client at the URL below:

```
https://raw.githubusercontent.com/Btibert3/datasets/master/phone-reviews1k.csv
```

In [ ]:
# imports
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics

In [ ]:
# bring in the data
df = pd.read_csv("https://raw.githubusercontent.com/Btibert3/datasets/master/phone-reviews1k.csv")

In [ ]:
# shape
df.shape

(1000, 2)

In [ ]:
# sample
df.sample(3)

reviews  ratings
851  Really its awesome with this price its a flats...        5
7    Am not of a Iphone user being this my first Ip...        4
857  a specific handy phone...with good features an...        5

In [ ]:
# data cleaning
df.dropna(inplace=True)

In [ ]:
# tokenize the text
cv = CountVectorizer(max_features=30)
dtm = cv.fit_transform(df.reviews).toarray()

In [ ]:
# make a dataset for the model
df2 = pd.DataFrame(dtm, columns=cv.get_feature_names_out())


In [ ]:
# quick confirmation
df2.sample(3)

and  apple  as  awesome  battery  best  ...  the  this  to  very  with  you
542    0      0   0        0        0     0  ...    0     0   1     0     0    0
867    3      0   0        0        0     0  ...    2     2   5     0     0    0
815    0      1   0        0        0     0  ...    1     0   0     0     0    1

[3 rows x 30 columns]

In [ ]:
# fit the model
tree = DecisionTreeClassifier(min_samples_leaf=1000, min_samples_split=2000, random_state=820)
tree.fit(df2, df.ratings)

DecisionTreeClassifier(min_samples_leaf=1000, min_samples_split=2000,
                       random_state=820)

In [ ]:
# how well do the model do?
cr = metrics.classification_report(df.ratings, tree.predict(df2))
print(cr)

              precision    recall  f1-score   support

           1       0.00      0.00      0.00        41
           2       0.00      0.00      0.00        19
           3       0.00      0.00      0.00        56
           4       0.00      0.00      0.00       189
           5       0.69      1.00      0.82       695

    accuracy                           0.69      1000
   macro avg       0.14      0.20      0.16      1000
weighted avg       0.48      0.69      0.57      1000



/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


 - 69% accurate
 - perfect recall on the top rating of 5!


 > My opinion is that we can work with the client to deploy the sampling strategy via the app, as we can predict the ratings well enough when there isn't one supplied.